### DISCUSSION TITLE
基于Metric数据模型实现influxdb协议的数据传输

### DISCUSSION THREAD

**Discussion by urnotsally** [alibaba/ilogtail#575] [Ideas]
# 背景
  基于Metric数据模型，实现对influxdb协议的支持。主要用于解决server端部署，需要接收多个客户端采集的指标，并完成数据上报的使用场景。

# 方案
## 1.Pipeline
![image](https://user-images.githubusercontent.com/30822030/207341210-c98495bc-7cfc-47eb-8279-600935209117.png)

### 2.1 [Input] service_http_server
<meta charset="utf-8"><div data-page-id="LqKedeBAzosjddxWxZPcsLH4n3T" data-docx-has-block-data="true"><ul start="1" class="list-bullet1"><li style="white-space:pre;" class="ace-line ace-line old-record-id-XcsMdAak0oIgMsxeOjacZtevnOd" data-list="bullet"><div>增加V2实现</div></li></ul><pre style="white-space:pre;" class="ace-line ace-line old-record-id-Z8QIdE8u0ook6sx89jmcOxOSnwd"><code class="language-Go" data-wrap="false"><div>type ServiceInputV2 interface {
   ServiceInput
   // StartService starts the ServiceInput's service, whatever that may be
   <span style="background-color:rgba(255,246,122,0.8)">StartService(PipelineContext) </span><span style="background-color:rgba(255,246,122,0.8)">error</span>
}</div></code></pre><ul start="1" class="list-bullet1"><li style="white-space:pre;" class="ace-line ace-line old-record-id-XAe8dcoUSoWGsqxWWATcOAOOnMg" data-list="bullet"><div>增加DecoderV2接口</div></li></ul><div style="white-space:pre;" class="ace-line ace-line old-record-id-TyoAdu08GoKkGAx6DqIc6njsnHe">&nbsp;&nbsp;用于将不同协议的输入数据解析到Metric数据模型</div><pre style="white-space:pre;" class="ace-line ace-line old-record-id-SEU0daUm6oecsExMpAocbJmAngc"><code class="language-Go" data-wrap="false"><div>// DecoderV2 used to parse buffer to pipeline events
type DecoderV2 interface {
    //Decode reader to groupInfo and events
    DecodeV2(data []byte, req *http.Request) (groupInfo *models.GroupInfo, events []*models.PipelineEvent, err error)
    ParseRequest(res http.ResponseWriter, req *http.Request, maxBodySize int64) (data []byte, statusCode int, err error)
}</div></code></pre><ul start="1" class="list-bullet1"><li style="white-space:pre;" class="ace-line ace-line old-record-id-LW2odGOUYoASWyxoNmqcKjnOnMe" data-list="bullet"><div>增加RawDecoder， 实现DecoderV2接口</div></li></ul><div style="white-space:pre;" class="ace-line ace-line old-record-id-Bi4Qd08c2oWAqwx8Xjnc3jkgnne">&nbsp;&nbsp;用于将HTTP数据解析到PipelineEvent数据结构，其中HTTP数据的Body以[]Byte数据保存在Event中；通过配置可以指定Query参数或者Header参数用于定义GroupInfo的metadata。</div><ul start="1" class="list-bullet1"><li style="white-space:pre;" class="ace-line ace-line old-record-id-VaQ4dAS04oQs4gxAHIqcUldonne" data-list="bullet"><div>新增配置项</div></li></ul><div>

配置项 | 类型 | 是否必选 | 说明
-- | -- | -- | --
Format | string | 是 | 新增"raw"选项，用于解析数据到ByteEvents
QueryParams | []string | 否 | 指定需要解析的Query参数，解析结果会以KeyValue放在GroupInfo.Metadata中，如["db"]。默认取值为空，即不解析。
QueryParamPrefix | string | 否 | 解析Query参数时需要添加的key前缀，如`_query_param_`。前缀会直接拼接在每个QueryParam前，无额外连接符，默认取值为空，即不增加前缀。
HeaderParams | []string | 否 | 指定需要解析的Header参数，解析结果会以KeyValue放在GroupInfo.Metadata中，如["Host"]。默认取值为空。
HeaderParamPrefix | string | 否 | 解析Header参数时需要添加的key前缀，如`_header_param_`。默认取值为空。
DisableUncompress | Boolean | 否 | 禁用对于请求数据的解压缩, 默认取值为:`false`<p>目前仅针对Raw Format有效</p><p>仅v2版本有效</p> 

</div></div>

### 2.2 [Aggregator] *aggregator_metadata_group
<meta charset="utf-8"><div data-page-id="LqKedeBAzosjddxWxZPcsLH4n3T" data-docx-has-block-data="true"><div style="white-space:pre;" class="ace-line ace-line old-record-id-AaIadSg82oYcS8xwAT9cxASGn2t">&nbsp;&nbsp;新增aggregator插件aggregator_byte_event_meta，实现AggregatorV2接口，用于GroupMetadata聚合，按照配置指定的keys聚合多个PipelineGroupEvents。</div><pre style="white-space:pre;" class="ace-line ace-line old-record-id-SIi2diYGaoG8sox4sBLcZnCwnRb"><code class="language-Go" data-wrap="false"><div>// AggregatorV2
// Apply, Push, and Reset can not be called concurrently, so locking is not
// required when implementing an Aggregator plugin.
type AggregatorV2 interface {
   Aggregator
   // Add the metric to the aggregator.
   <span style="background-color:rgba(255,246,122,0.8)">Record(*models.PipelineGroupEvents, PipelineContext) </span><span style="background-color:rgba(255,246,122,0.8)">error</span>
   // GetResult the current aggregates to the accumulator.
   <span style="background-color:rgba(255,246,122,0.8)">GetResult(PipelineContext) </span><span style="background-color:rgba(255,246,122,0.8)">error</span>
}</div></code></pre><div>

配置项 | 类型 | 是否必选 | 说明
-- | -- | -- | --
Type | string | 是 | 插件类型，固定为aggregator_metadata_group
GroupMetadataKeys | []string | 是 | 需要按照分组的Metadata key列表。如[`_query_param_db`]
GroupMaxEventLength | int | 否 | 聚合时，单个GroupEvent中Events数组的最大长度限制，默认长度为1024。
GroupMaxByteLength | int | 否 | 聚合时，单个GroupEvent中Events的最大字节长度限制，单位Byte，默认3MiB。该配置只针对ByteEvent数据有效

</div></div>

### 2.3 [Flusher] flusher_http
<meta charset="utf-8"><div data-page-id="LqKedeBAzosjddxWxZPcsLH4n3T" data-docx-has-block-data="true"><ul start="1" class="list-bullet1"><li style="white-space:pre;" class="ace-line ace-line old-record-id-SwocdcEYIoKMA6xEjxdcdc5sn3c" data-list="bullet"><div>增加FlusherV2实现</div></li></ul><pre style="white-space:pre;" class="ace-line ace-line old-record-id-SUs0dqIU2oIKGOxacftcCbgJn8c"><code class="language-Go" data-wrap="false"><div>type FlusherV2 interface {
   Flusher
   // Export data to destination, such as gRPC, console, file, etc.
   // It is expected to return no error at most time because IsReady will be called
   // before it to make sure there is space for next data.
   <span style="background-color:rgba(255,246,122,0.8)">Export([]*models.PipelineGroupEvents, PipelineContext) </span><span style="background-color:rgba(255,246,122,0.8)">error</span>
}</div></code></pre><div style="white-space:pre;" class="ace-line ace-line old-record-id-ICumd0iAQosQuUx76sbcv8Gwno9">&nbsp;&nbsp;新增支持配置：</div><div>

配置项 | 类型 | 是否必选 | 说明
-- | -- | -- | --
Query | Map<String,String> | 否 | 需要附加的请求参数。支持变量写法，动态从数据中提取*新增支持从Group.Metadata中提取。默认取值为空。
Convert.Protocol | string | 否 | logtail数据转换协议。*新增支持"raw"选项， 用于支持[]byte类型的Event直接输出到request body。<p>v2版本当前仅支持`raw`选项</p><p>默认取值为`custom_single`</p>
Convert.Encoding | string | 否 | ilogtail flusher数据转换编码。"Custom"<p>v2版本当前仅支持`Custom`选项</p><p>默认取值为`json`</p>

</div></div>

**Comment by urnotsally** [comment]
新增`inputserver`的配置参数`DisableUncompress `，用于配置是否针对header某些字段完成对请求体的解压缩，如 `Content-Encoding` 为 `gzip` 或者 `snappy`。
`DisableUncompress `配置默认值为`false`， 即识别header字段，若存在相应的`Content-Encoding`关键字，会先对Request Body进行解压，然后放入ByteEvents向后传递； 否则，将未解压的原始Body直接放入ByteEvents向后传递

**Comment by GledsonBastos** [comment]
Visualized 🧐

Em qua, 11 de jan de 2023 12:19, urnotsally ***@***.***>
escreveu:

**Comment by yyuuttaaoo** [comment]
文档中的问题：
1. HeaderParamPrefix 重复了，没说明白prefex和value的连接符是什么，或者可以配？
2. 整片文档的表格中，各个参数的默认值是什么？
3. 如["_query_param_db"]，是不是多了开头的_?
4. 聚合时，单个GroupEvent中Events的最大字节长度限制，单位Byte，默认2MiB。该配置只针对ByteEvent数据有效。2MiB是否有依据？我觉得可以从几方面考虑：1. 网络，万兆网卡，1000MB/s，就算2MB对应2ms，千兆网络就是20ms。2. 客户端内存考虑，比如压缩解压都需要单独申请内存 3. 从服务端考虑，能接受包的大小限制（SLS是5MB）。v1版本是3MiB，没特殊原因我觉得可以先统一。
5. flusher_http有动态query的话，需要动态headers吗？

**Reply by urnotsally** [reply]
问题回复：
1. 重复字段已删除。prefix和value直接连接，不设置额外连接符，在表格里已增加说明
2. 已补全非必填参数的默认值说明
3. 如["query_param_db"]的举例，是针对配置了prefix的情况，如果觉得有歧义，可以去掉
4. 可以统一为3MiB
5. flusher_http有动态query，v2版本支持从metadata中获取数据； 当前对于header暂时没有动态的需求，先不增加动态headers
